File: src/Alma.Abstraction/Configuration/Module/ModuleConfiguration.cs
namespace Alma.Configuration.Module;

public class ModuleConfiguration
{
    public string? Target { get; set; }
    public Dictionary<string, string>? Links { get; set; }
    public List<string>? Exclude { get; set; }
    public bool ExcludeReadme { get; set; } = true;

    public string? Install { get; set; }
    public string? Configure { get; set; }

    public ModuleConfiguration(string? target, Dictionary<string, string>? links, string? install, string? configure)
    {
        Target = target;
        Links = links;
        Install = install;
        Configure = configure;
    }

    public ModuleConfiguration Merge(ModuleConfiguration merge)
    {
        var mergedLinks = (Links ?? new Dictionary<string, string>())
            .Concat(merge.Links ?? new Dictionary<string, string>());
        return new ModuleConfiguration(
            merge.Target ?? Target,
            new Dictionary<string, string>(mergedLinks),
            merge.Install ?? Install,
            merge.Configure ?? Configure
        );
    }

    public static ModuleConfiguration Empty() =>
        new(null, new Dictionary<string, string>(), null, null);
}

File: src/config/moduleConfiguration.go
package config

import (
	"alma/helpers"
	"cmp"
	"encoding/json"
	"io"
	"os"
	"slices"
)

type ModuleConfiguration struct {
	Target        string            `json:"target"`
	Links         map[string]string `json:"links"`
	Exclude       []string          `json:"exclude"`
	ExcludeReadme bool              `json:"excludeReadme"`
	Install       string            `json:"install"`
	Configure     string            `json:"configure"`
}

func LoadModuleConfiguration(moduleConfigPath string) *ModuleConfiguration {
	jsonFile, err := os.Open(moduleConfigPath)

	if err != nil {
		return nil
	}

	defer jsonFile.Close()

	byteValue, _ := io.ReadAll(jsonFile)
	var moduleConfigurationRoot map[string]*ModuleConfiguration
	err = json.Unmarshal(byteValue, &moduleConfigurationRoot)
	if err != nil {
		return nil
	}

	platformKeys := make([]string, 0, len(moduleConfigurationRoot))

	for key := range moduleConfigurationRoot {
		if key != "default" && helpers.IsOnPlatform(key) {
			platformKeys = append(platformKeys, key)
		}
	}

	slices.SortFunc(platformKeys, func(i, j string) int {
		return cmp.Compare(i, j)
	})

	var moduleConfiguration *ModuleConfiguration
	if moduleConfigurationRoot["default"] != nil {
		moduleConfiguration = moduleConfigurationRoot["default"]
	}

	for _, platformKey := range platformKeys {
		moduleConfiguration.Merge(moduleConfigurationRoot[platformKey])
	}

	return moduleConfiguration
}

func (moduleConfig *ModuleConfiguration) Merge(mergeConfig *ModuleConfiguration) {
	// Merge links
	mergedLinks := make(map[string]string, len(moduleConfig.Links)+len(mergeConfig.Links))
	for key, value := range moduleConfig.Links {
		mergedLinks[key] = value
	}

	for key, value := range mergeConfig.Links {
		mergedLinks[key] = value
	}

	moduleConfig.Links = mergedLinks

	// Merge excludes
	mergedExcludes := make([]string, len(moduleConfig.Exclude)+len(mergeConfig.Exclude))
	excludeIndex := 0
	for _, value := range moduleConfig.Exclude {
		mergedExcludes[excludeIndex] = value
		excludeIndex++
	}

	for _, value := range mergeConfig.Links {
		mergedExcludes[excludeIndex] = value
		excludeIndex++
	}

	moduleConfig.Exclude = mergedExcludes

	// Simple properties
	if mergeConfig.Target != "" {
		moduleConfig.Target = mergeConfig.Target
	}
	if mergeConfig.Install != "" {
		moduleConfig.Install = mergeConfig.Install
	}
	if mergeConfig.Configure != "" {
		moduleConfig.Configure = mergeConfig.Configure
	}

	moduleConfig.ExcludeReadme = moduleConfig.ExcludeReadme || mergeConfig.ExcludeReadme
}
